test/e2e_federation/util.go

```diff
@@ -150,28 +139,104 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+		// if it's the case, delete them.
+		needToBeDeleted := false
+		err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+			var err error
+			needToBeDeleted, err = hasService(clientset, namespace, name)
+			if needToBeDeleted {
+				return true, nil
+			}
+			if err != nil && errors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, err
+		})
+
+		if err != nil {
+			framework.Failf("Getting the service %q creation status after a partial createService(): %v", service.Name, err)
+		}
+
+		if needToBeDeleted {
+			if err = deleteService(clientset, namespace, name, nil); err != nil {
+				framework.ExpectNoError(err, "Deleting service %q after a partial createService() error", service.Name)
+				return nil, err
+			}
+		}
+
+		// creation failed, lets try with another port
+		// first remove from the availablePorts the port with which the creation failed
+		availablePorts = append(availablePorts[:i], availablePorts[i+1:]...)
+	}
+
+	return nil, err
+}
+
 func createServiceOrFail(clientset *fedclientset.Clientset, namespace, name string) *v1.Service {
 	service, err := createService(clientset, namespace, name)
 	framework.ExpectNoError(err, "Creating service %q in namespace %q", service.Name, namespace)
 	By(fmt.Sprintf("Successfully created federated service %q in namespace %q", name, namespace))
 	return service
 }
 
-func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, serviceName string, orphanDependents *bool) {
-	if clientset == nil || len(namespace) == 0 || len(serviceName) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteServiceOrFail: clientset: %v, namespace: %v, service: %v", clientset, namespace, serviceName))
+func hasService(clientset *fedclientset.Clientset, namespace string, serviceName string) (bool, error) {
+	_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
+	if err == nil {
+		// it means that the Get request found the service
+		return true, nil
 	}
-	framework.Logf("Deleting service %q in namespace %v", serviceName, namespace)
+	// it means that the service didn't exist or another possible error request (internal, timeout) that needs to be check
+	return false, err
+}
+
+func deleteService(clientset *fedclientset.Clientset, namespace string, serviceName string, orphanDependents *bool) error {
 	err := clientset.Services(namespace).Delete(serviceName, &metav1.DeleteOptions{OrphanDependents: orphanDependents})
-	framework.ExpectNoError(err, "Error deleting service %q from namespace %q", serviceName, namespace)
+	if err != nil {
+		return err
+	}
 	// Wait for the service to be deleted.
 	err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
-		_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
-		if err != nil && errors.IsNotFound(err) {
+		found, err := hasService(clientset, namespace, serviceName)
```

Comment:
I don't find any significant value in hasService function and it has made the logic little complex. Can we get rid of this function and use `clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})` as earlier.